src/content/developers/docs/scaling/zk-rollups/index.md

```diff
@@ -1,34 +1,206 @@
+
+Besides, users cannot interact with the rollup if operators don't publish data on-chain. An inability to access state data means users cannot know the state of their accounts or initiate transactions (e.g., withdrawals) based on state updates.  
+
+### Transaction finality {#transaction-finality}
+
+Ethereum acts as a settlement layer for ZK-rollups: L2 transactions are finalized only if the L1 contract accepts the validity proof. This eliminates the risk of malicious operators corrupting the chain (e.g., stealing rollup funds) since every transaction must be approved on Mainnet. Also, Ethereum guarantees that user operations cannot be reversed once finalized on L1.  
+
+### Censorship resistance {#censorship-resistance}
+
+Most ZK-rollups use a "supernode" (the operator) to execute transactions, produce batches, and submit blocks to L1. While this ensures efficiency, it increases the risk of censorship: malicious ZK-rollup operators can censor users by refusing to include their transactions in batches. 
+
+As a security measure, ZK-rollups allow users to submit transactions directly to the rollup contract on Mainnet if they think they are being censored by the operator. This allows users to force an exit from the ZK-rollup to Ethereum without having to rely on the operator’s permission.
+
+## How do zk-rollups work? {#how-do-zk-rollups-work}
+
+### Transactions {#transactions}
+
+Users in the ZK-rollup sign transactions and submit to L2 operators for processing and inclusion in the next batch. In some cases, the operator is a centralized entity, called a sequencer, who executes transactions, aggregates them into batches, and submits to L1. The sequencer in this system is the only entity allowed to produce L2 blocks and add rollup transactions to the ZK-rollup contract.  
+
+Other ZK-rollups may rotate the operator role by using a [proof-of-stake](/developers/docs/consensus-mechanisms/pos/) validator set. Prospective operators deposit funds in the rollup contract, with the size of each stake influencing the staker’s chances of getting selected to produce the next rollup batch. The operator’s stake can be slashed if they act maliciously, which incentivizes them to post valid blocks.  
+
+#### How ZK-rollups publish transaction data on Ethereum {#how-zk-rollups-publish-transaction-data-on-ethereum} 
+
+As explained, transaction data is published on Ethereum as `calldata`. Calldata is a data area in a smart contract used to pass arguments to a function. Although `calldata` is similar to [memory](/developers/docs/smart-contracts/anatomy/#memory), it isn’t stored as part of Ethereum’s state; rather, it persists on-chain as part of the Ethereum chain's [history logs](https://docs.soliditylang.org/en/latest/introduction-to-smart-contracts.html?highlight=memory#logs). This makes it a cheaper alternative to `memory` or `storage` for storing data on-chain. 
```

Comment:
See earlier comments- `calldata` has been discussed multiple times already and is only now defined.

Reply:
This is similar to what I said in the earlier comments: I didn't see a way of properly explaining some concepts like calldata at the beginning without taking a massive detour. The idea was to assume readers (who are likely developers) already know what it meant, and if they didn't, I could explain it properly in the right context. 

For example, this sentence "ZK-rollups write transactions to Ethereum as calldata. `calldata` is where data that is included in external calls to smart contract functions gets stored" may not make an impression on a reader at the beginning since we haven't even explained that the rollup is managed by an L1 contract. 

Going into `calldata` at the transaction phase made sense since readers are already aware of two facts: a. ZK-rollups rely on L1 contracts b. Operators publish data to this contract. Then they can tie both facts in their minds and understand what `calldata` is and why it's necessary.